File: frontend/app/utils/interface/file.model.ts
import * as tus from "tus-js-client";

export enum FileStatus {
  STARTED = 1,
  FAILED = 2,
  CANCELED = 3,
  COMPLETED = 4,
}

export interface FileAccessModel {
  is_public: boolean;
  shared_with: number[];
  url?: string;
}

export interface FileListingModel {
  id: number;
  name: string;
  size: number;
  owner_name: string;
  upload_status: FileStatus;
  access: FileAccessModel;
  created_on: Date;
  updated_on: Date;
  is_deleted: boolean;
}

export interface FileListingResponseModel {
  data: FileListingModel[];
  uploadData: FileUploadModel[];
  count: number;
}

export interface FileUploadModel {
  file_id: number;
  progress: number;
  status: FileStatus;
  url?: string;
  tusClient?: tus.Upload;
}

File: backend/app/model/file_model.py
from datetime import datetime
from typing import List
from pydantic import BaseModel

from ..model.shared_enum import FileUploadEnum


class CreateFileModel(BaseModel):
    name: str
    size: float


class UpdateFileStatusModel(BaseModel):
    file_id: int
    status: FileUploadEnum
    url: str | None = None


class FileAccessModel(BaseModel):
    is_public: bool
    shared_with: List[int]
    url: str | None = None


class FileUploadModel(BaseModel):
    file_id: int
    progress: float
    status: FileUploadEnum
    url: str | None = None


class FileDetailsModel(BaseModel):
    id: int
    name: str
    size: float
    owner_name: str
    upload_status: FileUploadEnum
    access: FileAccessModel
    created_on: datetime
    updated_on: datetime
    is_deleted: bool


class FileListingResponseModel(BaseModel):
    data: List[FileDetailsModel]
    uploadData: List[FileUploadModel]
    count: int
